### ISSUE TITLE
TRACE reproduction: early-stopping mechanism and exact sparsity ratio unclear in released code

### ISSUE THREAD

**Issue by Peiyu-Georgia-Li** [JinhuiBot/PaRSP#1]
Hi, thanks for releasing the code for PaRSP — really helpful for reproducing the TRACE results. Two things I ran into while trying to match the paper's TRACE setting exactly:

**1. Early stopping mentioned in the paper isn't in the released training code**

The paper states:

_"A consistent early stopping mechanism was applied across all tasks to prevent overfitting, with the maximum training duration capped at specific epochs depending on the benchmark complexity."_

and for TRACE specifically:

_"we maintained the ~17% activation ratio but increased the maximum training duration to 5 epochs to ensure sufficient adaptation."_

This reads as: 5 epochs is an upper bound, with some early-stopping criterion that can stop training earlier per task. However, in parsp/engine/trainer_llama.py (and the equivalent trainer_qwen.py), Trainer is constructed without an eval_dataset:

trainer = Trainer(
    model=model,
    args=script_args,
    train_dataset=train_dataset,
    tokenizer=tokenizer,
    data_collator=DataCollatorForSeq2Seq(tokenizer=tokenizer, model=model, padding="longest"),
)
trainer.train()
Without an eval_dataset, HF Trainer has nothing to evaluate against during training, so it can't implement early stopping. I couldn't find EarlyStoppingCallback, patience, load_best_model_at_end, or evaluation_strategy anywhere in the repo, and shells/train_llama.sh / shells/train_qwen.sh just pass a fixed --num_train_epochs. So as released, training runs for exactly num_train_epochs every time, not "up to N with early stopping."

Question: Is the early-stopping logic simply not included in this release, or am I missing how it's wired in? If it's not included, what was the actual stopping criterion for the reported TRACE numbers (validation loss patience, fixed step count, etc.)?

**2. Exact sparsity ratio for TRACE is ambiguous**

The paper says TRACE uses "~17% activation ratio," but I can't find a TRACE-specific config anywhere in the repo — configs/build_arguments.py defines layer_ratio default 0.38 and neuron_ratio default 0.5, while the only example script, shells/train_llama.sh, overrides neuron_ratio to 0.4 (for the Standard-CL/dbpedia setting, not TRACE) and leaves layer_ratio at its default 0.38. There's no shells/train_*_trace.sh or similar.

Question: What exact layer_ratio/neuron_ratio combination (or other knobs) was used to produce the "~17%" figure for TRACE specifically? Is layer_ratio=0.38, neuron_ratio=0.4 from the example script the right one to use, or was a different combination used for TRACE vs. the Standard-CL benchmark?

Thanks for any clarification. I am trying to get results that match the paper as closely as possible.
